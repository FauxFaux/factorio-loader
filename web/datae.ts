import { BlockContent } from '../scripts/load-recs';
import { FRecipe, JFluid, JItem, JRecipe } from './objects';
import { Colon } from './muffler/colon';
import { LtnSummary, precomputeLtnSummary } from './ltn-summary';
import { makeUpRecipe } from './muffler/walk-recipes';

export type Coord = readonly [number, number];
// concatenated positions; i.e. `${x},${y}`
export type ConcPos = string;

// e.g. `automated-factory`
export type FactoryClass = string;
// e.g. `automated-factory-mk01`
export type Factory = string;

export type Limitation = string;

export const data = {
  doc: {} as Record<string, BlockContent>,
  meta: {} as {
    factories: Record<
      FactoryClass,
      Record<
        Factory,
        { speed: number; modules: number; dims: [number, number] }
      >
    >;
    modules: Record<Limitation, Record<string, number>>;
    isSpawn: Loc[];
  },
  /** @deprecated replace with colons */
  items: {} as Record<string, JItem>,
  icons: {} as Record<string, string>,
  /** @deprecated replace with colons */
  fluids: {} as Record<string, JFluid>,
  recipes: {} as {
    regular: Record<string, JRecipe>;
    voidableItems: string[];
    barrelFormOf: Record<string, string>;
    placeOverrides: Record<string, string>;
  },
  prodStats: {} as Record<
    Colon,
    { input?: FlowStats; output?: FlowStats; ltn?: number }
  >,
  trainPulses: {} as { byColon: Record<Colon, Pulse[]> },
  flowDiagrams: [] as string[],
  technologies: {} as Record<
    string,
    {
      researched: boolean;
      requires: string[];
      unlocks: string[];
    }
  >,
  cp: {} as {
    ticks: number[];
    byPos: Record<ConcPos, { runs?: number[]; recipe?: string; name?: string }>;
  },
  maps: {} as {
    maps: MapRef[];
  },
} as const;

export interface FlowStats {
  /** total ever made/consumed */
  total: number;
  /** 5s, ... 100h */
  perTime: number[];
}

export const computed = {
  ltnSummary: {} as Record<string, LtnSummary>,
  /** fluid -> item ({ "fetal-serum": "fetal-serum-barrel" }) */
  fluidBarrel: {} as Record<string, string>,
  /** item -> fluid ({ "fetal-serum-barrel": "fetal-serum" }) */
  barrelFluid: {} as Record<string, string>,
  recipeExecs: {} as Record<string, number>,
  recipes: [] as FRecipe[],
};

export type Loc = string;
/** from, to, start, duration, amount */
export type Pulse = [Loc, Loc, number, number, number];

export interface MapRef {
  date: string;
  tick: number;
  speed: number;
  trains: number;
  researchName: string;
  researchProgress: number;
  hasMap: boolean;
}

export function precompute() {
  computed.ltnSummary = precomputeLtnSummary();
  computed.fluidBarrel = data.recipes.barrelFormOf;
  computed.barrelFluid = Object.fromEntries(
    Object.entries(computed.fluidBarrel).map(([k, v]) => [v, k] as const),
  );

  for (const dat of Object.values(data.cp.byPos)) {
    if (!(dat.recipe && dat.runs)) {
      continue;
    }
    computed.recipeExecs[dat.recipe] =
      (computed.recipeExecs[dat.recipe] ?? 0) +
      (dat.runs.reduce((a, b) => a + b, 0) ?? 0);
  }

  computed.recipes = Object.keys(data.recipes.regular)
    .map((name) => makeUpRecipe(name))
    .filter((r): r is FRecipe => !!r);
}
