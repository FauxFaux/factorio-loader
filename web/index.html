<html lang="en">
<head><title></title></head>
<body>
<div id="app">Waiting for JS to boot</div>
<script src="bundle.js"></script>
<script>
  app.init(document.getElementById('app'));
</script>
</body>
</html>
