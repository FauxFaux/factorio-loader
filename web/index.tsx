import 'preact/debug';
import { Component, render } from 'preact';

import * as docData from '../data/data.json';
import { BlockContent } from '../scripts/load-recs';
const doc = docData as Record<string, BlockContent>;

class App extends Component {
  render() {
    const blocks = [];

    for (const [loc, obj] of Object.entries(doc)) {
      blocks.push(
        <h1>
          <a name="${loc}" href="#${loc}">
            ${loc}
          </a>
        </h1>,
      );
      const list = [];
      if (obj.tags.length) {
        list.push(<li>Tags: ${obj.tags.sort().join(', ')}</li>);
      }
      if (Object.keys(obj.asm).length) {
        const asmList = [];
        for (const [label, count] of Object.entries(obj.asm).sort(
          ([, a], [, b]) => b - a,
        )) {
          asmList.push(
            <li>
              ${count} * ${label}
            </li>,
          );
        }

        list.push(
          <li>
            Assemblers: <ul>${asmList}</ul>
          </li>,
        );
      }

      if (obj.stop.length) {
        list.push(
          <>
            <li>Train stops</li>
            <ul>
              {obj.stop.map((stop) => {
                const nonVirt = stop.items
                  .sort(([, , a], [, , b]) => Math.abs(b) - Math.abs(a))
                  .filter(([kind]) => kind !== 'virtual');
                return (
                  <li>
                    ${stop.name}
                    <ul>
                      {nonVirt.map(([kind, name, count]) => (
                        <li>
                          ${count} * ${kind}:${name}
                        </li>
                      ))}
                    </ul>
                  </li>
                );
              })}
            </ul>
          </>,
        );
      }
      blocks.push(list);
    }
    return (
      <ul>{blocks}</ul>
    );
  }
}

export function init(element: HTMLElement) {
  element.innerHTML = '';
  render(<App />, element);
}
