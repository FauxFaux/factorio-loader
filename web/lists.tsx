import { Component, ComponentChild } from 'preact';
import { data } from './datae';
import { Stop } from '../scripts/load-recs';
import { ItemOrFluid } from './objects';
import { compareWithoutIcons } from './muffler/names';
import { Colon, tupleToColon } from './muffler/colon';

function smatch(name: string, props: { search?: string }) {
  // TODO: proper string comparison
  if (!props.search) return true;
  return name.toLowerCase().includes(props.search.toLowerCase());
}

export class StationList extends Component<
  { limit?: number },
  { search?: string }
> {
  stations: Record<string, { blockNo: string; stop: Stop }[]> = {};

  constructor() {
    super();
    for (const [blockNo, { stop: stops }] of Object.entries(data.doc)) {
      for (const stop of stops) {
        if (!this.stations[stop.name]) this.stations[stop.name] = [];
        this.stations[stop.name].push({ blockNo, stop });
      }
    }
  }

  onInput = (e: any) => {
    const search = e.target.value;
    this.setState({ search });
  };

  render(
    state: { limit?: number },
    props: { search?: string },
  ): ComponentChild {
    const found = Object.entries(this.stations)
      .filter(([name]) => {
        if (!props.search) return true;
        // TODO: proper string comparison
        // TODO: search other attributes
        return smatch(name, props);
      })
      .sort(([a], [b]) => compareWithoutIcons(a, b));
    return (
      <div>
        <p>
          <input
            type="text"
            onInput={this.onInput}
            class="form-control"
            placeholder="Search station names..."
          ></input>
        </p>
        <ul style={'list-style: none; padding-left: 0'}>
          {found.slice(0, state.limit ?? Infinity).map(([name, stops]) => (
            <li>
              <GpsLink caption={name} gps={stops[0].stop.gps} />{' '}
              <RenderIcons text={name} />{' '}
              {stops.length !== 1 ? `(${stops.length} stops)` : ''}
              {stops.map(({ blockNo }) => (
                <a style="padding: 0 4px" href={`/block/${blockNo}`}>
                  {blockNo}
                </a>
              ))}
            </li>
          ))}
          {found.length >= (state.limit ?? Infinity) ? <li>...</li> : <></>}
        </ul>
      </div>
    );
  }
}

export class ItemIcon extends Component<{ name: string; alt: string }> {
  render(props: { name: string; alt: string }): ComponentChild {
    return (
      <span
        className="icon-sprite"
        title={props.alt}
        style={`background: url("../data/icons.png") ${data.icons[props.name]}`}
      />
    );
  }
}

export class RenderIcons extends Component<{ text: string }> {
  render(props: { text: string }): ComponentChild {
    const text = props.text;
    if (!text.includes('[')) return text;
    const parts: ComponentChild[] = [];
    let prev = 0;
    for (const ma of text.matchAll(/\[([a-z0-9-]+)=([a-z0-9-]+)]/g)) {
      parts.push(' ' + text.slice(prev, ma.index) + ' ');
      prev = ma.index! + ma[0].length;
      const type = ma[1];
      const name = ma[2];
      if (data.icons[name]) {
        parts.push(
          <ItemIcon
            name={name}
            alt={`${type} ${name} (${
              (data as any)[`${type}s`]?.[name]?.localised_name
            })`}
          />,
        );
      } else {
        parts.push(
          <i>
            [{type}={name}]
          </i>,
        );
      }
    }
    parts.push(text.slice(prev));
    return <>{parts}</>;
  }
}

interface ItemListProps {
  limit?: number;
  onPick?: (colon: Colon) => void;
}

export class ItemList extends Component<ItemListProps, { search?: string }> {
  itemAsms: Record<Colon, number> = {};

  constructor() {
    super();
    for (const block of Object.values(data.doc)) {
      for (const [label, { count }] of Object.entries(block.asm)) {
        const [, recipe] = label.split('\0');
        for (const product of data.recipes.regular[recipe]?.products ?? []) {
          const colon = product.colon;
          if (!this.itemAsms[colon]) this.itemAsms[colon] = 0;
          this.itemAsms[colon] += count;
        }
      }
    }
  }

  onInput = (e: any) => {
    const search = e.target.value;
    this.setState({ search });
  };

  render(props: ItemListProps, state: { search?: string }): ComponentChild {
    const found = [
      ...Object.entries(data.items).map(([n, i]) => [n, i, 'item'] as const),
      ...Object.entries(data.fluids).map(([n, i]) => [n, i, 'fluid'] as const),
    ]
      .filter(([name, item]) => {
        // TODO: search other attributes?
        return smatch(name, state) || smatch(item.localised_name, state);
      })
      .sort(([, { localised_name: a }], [, { localised_name: b }]) =>
        compareWithoutIcons(a, b),
      );
    return (
      <div>
        <p>
          <input
            type="text"
            onInput={this.onInput}
            class="form-control"
            placeholder="Search items and fluids..."
          ></input>
        </p>
        <table>
          {found.slice(0, props.limit ?? Infinity).map(([name, , type]) => (
            <tr class="item-list">
              {props.onPick ? (
                <td>
                  <button
                    class="btn btn-sm btn-outline-secondary"
                    onClick={() => props.onPick?.(tupleToColon([type, name]))}
                  >
                    +
                  </button>
                </td>
              ) : (
                <></>
              )}
              <td title="being made in x factories">
                {this.itemAsms[tupleToColon([type, name])]}
              </td>
              <td>
                <ItemIcon name={name} alt={name} />
              </td>
              <td>
                <ItemOrFluid type={type} name={name} />
              </td>
            </tr>
          ))}
          {found.length >= (props.limit ?? Infinity) ? (
            <tr>
              <td colSpan={4}>...</td>
            </tr>
          ) : (
            <></>
          )}
        </table>
      </div>
    );
  }
}

export class GpsLink extends Component<
  { caption: string; gps: readonly [number, number] },
  { copied?: boolean }
> {
  doCopy = async () => {
    try {
      const { caption, gps } = this.props;
      await navigator.clipboard.writeText(
        `[gps=${gps[0].toFixed()},${gps[1].toFixed()}] ${caption}`,
      );
      this.setState({ copied: true });
    } catch (err) {
      alert('Failed to copy to clipboard: ' + err);
    }
  };

  onLeave = () => {
    this.setState({ copied: false });
  };
  render(
    props: {
      caption: string;
      gps: readonly [number, number];
    },
    state: { copied?: boolean },
  ): ComponentChild {
    const check = require('svg-url-loader!./check.svg');
    const pin = require('svg-url-loader!./pin.svg');

    if (state.copied) {
      return (
        <span className={'gps-link-button'} onMouseLeave={this.onLeave}>
          <img src={check} alt={'done!'} />
        </span>
      );
    }
    const msg = 'copy location in chat format';
    return (
      <span className={'gps-link-button'} onClick={this.doCopy} title={msg}>
        <img src={pin} alt={msg} />
      </span>
    );
  }
}
