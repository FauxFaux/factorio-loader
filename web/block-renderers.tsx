import { Component } from 'preact';
import { BlockContent } from '../scripts/load-recs';
import { ColonJoined, Item, Recipe } from './objects';
import { GpsLink, RenderIcons } from './lists';
import { Coord, data } from './datae';
import {
  colonMapCombinator,
  colonMapItems,
  ltnMinTransfer,
  settingsMap,
} from './muffler/stations';
import { humanise } from './muffler/human';
import { compareWithoutIcons } from './muffler/names';
import { expActLtn, LtnFlow, LtnPercent } from './ltn-avail';

export function recipeDifference(brick: BlockContent) {
  const inputs: Set<string> = new Set();
  const outputs: Set<string> = new Set();
  for (const [label] of Object.entries(brick.asm)) {
    const [, recipe] = label.split('\0');
    const recp = data.recipes.regular[recipe];
    if (!recp) continue;

    for (const ing of recp.ingredients ?? []) {
      inputs.add(ing.colon);
    }

    for (const prod of recp.products ?? []) {
      outputs.add(prod.colon);
    }
  }

  if (brick.boilers > 0) {
    inputs.add('fluid:water');
    outputs.add('fluid:steam');
  }

  const wanted = [...inputs].filter((input) => !outputs.has(input));
  const exports = [...outputs].filter((output) => !inputs.has(output));

  return { wanted, exports };
}

export class Assemblers extends Component<{ brick: BlockContent }> {
  render(props: { brick: BlockContent }) {
    const sorted = Object.entries(props.brick.asm).sort(
      ([, { count: a }], [, { count: b }]) => b - a,
    );
    return (
      <>
        <ul>
          {sorted.map(([label, props]) => {
            const [, recipe] = label.split('\0');
            return (
              <li>
                <AssemblerCount label={label} props={props} />
                making <Recipe name={recipe} />
              </li>
            );
          })}
        </ul>
      </>
    );
  }
}

export const AssemblerCount = ({
  label,
  props,
}: {
  label: string;
  props: { count: number; locations: Coord[] };
}) => {
  const [machine, recipe] = label.split('\0');
  const machineName = data.items[machine]?.localised_name ?? machine;
  const recipeName = data.recipes.regular[recipe]?.localised_name ?? recipe;
  return (
    <>
      {props.locations.map((loc) => (
        <GpsLink caption={`a ${machineName} making ${recipeName}`} gps={loc} />
      ))}
      {props.count} &times; <Item name={machine} />{' '}
    </>
  );
};

export class TrainStops extends Component<{ stop: BlockContent['stop'] }> {
  render(props: { stop: BlockContent['stop'] }) {
    return (
      <table class="ltn-avail-block">
        {props.stop
          .sort((a, b) => compareWithoutIcons(a.name, b.name))
          .map((stop) => {
            const settings = settingsMap(stop);
            const items = colonMapItems(stop);
            const combinators = colonMapCombinator(stop);
            return (
              <>
                <tr>
                  <th colSpan={5}>
                    <b>
                      <GpsLink caption={stop.name} gps={stop.gps} />{' '}
                      <RenderIcons text={stop.name} />
                    </b>
                  </th>
                </tr>
                {Object.entries(items)
                  .sort(([a], [b]) => a.localeCompare(b))
                  .map(([colon, count]) => {
                    const req = combinators[colon];
                    if (req) {
                      const v = expActLtn(combinators, items, colon);
                      if (!v)
                        return (
                          <tr>
                            <td>ERROH {colon} isn't a combinator</td>
                          </tr>
                        );
                      const { actual, expected } = v;
                      return (
                        <tr>
                          <td title="import">➡</td>
                          <td> {humanise(actual)}</td>
                          <td>
                            <LtnPercent
                              actual={actual}
                              expected={expected}
                              decimate={true}
                            />
                          </td>
                          <td>
                            <LtnFlow
                              flows={{
                                totalFlow: data.prodStats[colon]?.ltn,
                                flow: stop.flowTo[colon],
                              }}
                            />
                          </td>
                          <td>
                            <ColonJoined colon={colon} />
                          </td>
                        </tr>
                      );
                    }
                    const expected = ltnMinTransfer(colon, settings);
                    return (
                      <tr>
                        <td title="export">️⬅</td>
                        <td>️{humanise(count)}</td>
                        <td>
                          <LtnPercent actual={count} expected={expected} />
                        </td>
                        <td>
                          <LtnFlow
                            flows={{
                              totalFlow: data.prodStats[colon]?.ltn,
                              flow: stop.flowFrom[colon],
                            }}
                          />
                        </td>
                        <td>
                          <ColonJoined colon={colon} />
                        </td>
                      </tr>
                    );
                  })}
              </>
            );
          })}
      </table>
    );
  }
}
