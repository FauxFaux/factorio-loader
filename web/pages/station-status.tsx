import { Component } from 'preact';

import { data } from '../datae.js';
import { Stop } from '../../scripts/load-recs.js';
import { GpsLink, RenderIcons } from '../lists.jsx';
import { ColonJoined } from '../objects.jsx';
import { tupleToColon } from '../muffler/colon.js';
import {
  colonMapItems,
  ltnMinTransfer,
  settingsMap,
  Stat,
  stations,
} from '../muffler/stations.js';

export class StationStatus extends Component {
  render() {
    const stops = stations();

    const providers: Stat[] = [];
    const requesters: Stat[] = [];
    const boring: Stat[] = [];
    const bugs: Stat[] = [];

    for (const [loc, stop] of stops) {
      if (
        /empty barrel/i.exec(stop.name) ||
        /storehouse provide/i.exec(stop.name) ||
        /passenger stop/.exec(stop.name) ||
        /virtual-signal=ltn-depot/.exec(stop.name)
      ) {
        boring.push([loc, stop]);
      } else if (stop.provides.length > 0) {
        providers.push([loc, stop]);
      } else if (/\b(?:Request|drop)\b/i.exec(stop.name)) {
        requesters.push([loc, stop]);
      } else {
        bugs.push([loc, stop]);
      }
    }

    const providerObjs = providers.map(([loc, stop]) => {
      const declaring = stop.provides;

      const settings = settingsMap(stop);

      const available = colonMapItems(stop);

      const health: Record<string, number> = {};

      for (const tuple of declaring) {
        const colon = tupleToColon(tuple);
        health[colon] =
          (available[colon] ?? 0) / ltnMinTransfer(colon, settings);
      }

      return { stop: [loc, stop], health, available, settings } as const;
    });

    return (
      <>
        <div className="row">
          <h2>Provider-like things</h2>
          <p>
            Congratulations on waiting for this page to load! This shows the{' '}
            <i>percentage satisfaction</i> of a train by a provider. If the
            stack size is 5, and the LTN stack requirement is 10, a train is
            assumed to want 50 items. If there are 40 items present, the
            satisfaction is 40/50 = 80%. If there are 500 items present, the
            satisfaction is 1,000%.
          </p>
          <table class="wtb">
            {providerObjs
              .sort(({ health: a }, { health: b }) => {
                return (
                  Math.min(...Object.values(a)) - Math.min(...Object.values(b))
                );
              })
              .map(({ stop, health, available, settings }) => (
                <tr>
                  <td>
                    <StopLine stop={stop} />
                  </td>
                  <td>
                    <table>
                      {Object.entries(health)
                        .sort(([, a], [, b]) => a - b)
                        .map(([colon, health]) => (
                          <tr>
                            <td>
                              {(health * 100).toLocaleString('en', {
                                maximumFractionDigits: 0,
                              })}
                              %
                            </td>
                            <td>
                              <ColonJoined colon={colon} />
                            </td>
                          </tr>
                        ))}
                    </table>
                  </td>
                  {/*{JSON.stringify(available)}*/}
                  {/*{JSON.stringify(settings)}*/}
                </tr>
              ))}
          </table>
        </div>
        <div class="row">
          <h2>Names not matching a known pattern</h2>
          <ul>
            {bugs.map((stop) => (
              <li>
                <StopLine stop={stop} />
              </li>
            ))}
          </ul>
        </div>
        <div class="row">
          <h2>Names deemed too boring to consider</h2>
          <ul>
            {boring.map((stop) => (
              <li>
                <StopLine stop={stop} />
              </li>
            ))}
          </ul>
        </div>
      </>
    );
  }
}

export class StopLine extends Component<{ stop: readonly [string, Stop] }> {
  render(props: { stop: [string, Stop] }) {
    const stop = props.stop[1];
    const loc = props.stop[0];
    return (
      <span>
        <GpsLink caption={stop.name} gps={stop.gps} />{' '}
        <RenderIcons text={stop.name} /> (<BlockLink loc={loc} />)
      </span>
    );
  }
}

export class BlockLink extends Component<{ loc: string }> {
  render(props: { loc: string }) {
    return (
      <a
        href={`/block/${props.loc}`}
        title={data.doc[props.loc].tags.sort().join(', ')}
      >
        {props.loc}
      </a>
    );
  }
}
