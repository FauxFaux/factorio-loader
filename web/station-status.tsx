import { Component } from 'preact';
import { data } from './index';
import { Stop } from '../scripts/load-recs';
import { cleanupName, ItemIcon, RenderIcons } from './lists';
import { singularize } from 'inflection';
import { Item } from './objects';

function strIEq(a: string, b: string): boolean {
  return (
    0 ===
    a.localeCompare(b, 'en', {
      sensitivity: 'base',
      usage: 'search',
      ignorePunctuation: true,
    })
  );
}

const labelLookupCache: Record<string, string> = {};

export function closelyMatchesItemName(label: string): string | null {
  if (Object.keys(labelLookupCache).length === 0) {
    for (const [name, item] of Object.entries(data.items)) {
      labelLookupCache[item.localised_name.toLowerCase()] = name;
    }
  }

  const quick = labelLookupCache[label.toLowerCase()];
  if (quick) return quick;

  switch (label.toLowerCase()) {
    // typo: inductor
    case 'air-core conductor':
      return 'inductor1';
  }
  for (const [name, item] of Object.entries(data.items)) {
    if (
      strIEq(label, item.localised_name) ||
      strIEq(singularize(label), item.localised_name)
    ) {
      return name;
    }
  }

  return null;
}

function closelyMatches(label: string): string | null {
  let cleaned = cleanupName(label);
  let v = closelyMatchesItemName(cleaned);
  if (v) return v;
  v = closelyMatchesItemName(cleaned + ' barrel');
  if (v) return v;
  if (cleaned.startsWith('Auog ')) {
    v = closelyMatchesItemName(cleaned.slice('Auog '.length));
    if (v) return v;
  }

  return null;
}

export function provideStationPurpose(name: string): Set<string> {
  const matches = new Set<string>();
  for (const ma of name.matchAll(/\[item=([a-z0-9-]+)]/g)) {
    const item = ma[1];
    if (item in data.items) {
      matches.add(item);
    }
  }

  if (matches.size) return matches;
  name = cleanupName(name);
  let ma = name.match(/^([a-zA-Z0-9 /-]+) Provide/i);
  if (ma) {
    for (const part of ma[1].split('/')) {
      const guess = closelyMatches(part);
      if (guess) {
        matches.add(guess);
      }
    }
  }
  return matches;
}

export class StationStatus extends Component {
  render() {
    const stops = Object.entries(data.doc).flatMap(([loc, brick]) =>
      brick.stop.map((stop) => [loc, stop] as const),
    );

    type Stat = [string, Stop];

    const providers: Stat[] = [];
    const requesters: Stat[] = [];
    const boring: Stat[] = [];
    const bugs: Stat[] = [];

    for (const [loc, stop] of stops) {
      if (
        /empty barrel/i.exec(stop.name) ||
        /storehouse provide/i.exec(stop.name) ||
        /passenger stop/.exec(stop.name) ||
        /virtual-signal=ltn-depot/.exec(stop.name)
      ) {
        boring.push([loc, stop]);
      } else if (/\bProvide\b/i.exec(stop.name)) {
        providers.push([loc, stop]);
      } else if (/\b(?:Request|drop)\b/i.exec(stop.name)) {
        requesters.push([loc, stop]);
      } else {
        bugs.push([loc, stop]);
      }
    }

    type Settings = Record<string, number>;

    const providerObjs: any[] = [];

    for (const [loc, stop] of providers) {
      const declaring = provideStationPurpose(stop.name);

      const settings: Settings = Object.fromEntries(
        stop.settings
          .filter(
            ([kind, name]) => kind === 'virtual' && name.startsWith('ltn-'),
          )
          .map(([, name, value]) => [name, value] as const),
      );

      const available = Object.fromEntries(
        stop.items
          .filter(([kind]) => kind === 'item')
          .map(([, name, value]) => [name, value] as const),
      );

      const health: Record<string, number> = {};

      for (const declared of declaring) {
        const item = data.items[declared];
        const expectedByStack =
          item.stack_size * (settings['ltn-provider-stack-threshold'] ?? 10);
        const expectedByCount = settings['ltn-provider-threshold'] ?? 1;
        const expected = Math.max(expectedByStack, expectedByCount);
        health[declared] = (available[declared] ?? 0) / expected;
      }

      providerObjs.push({ stop: [loc, stop], health, available, settings });
    }

    return (
      <>
        <div className="row">
          <h2>Provider-like things</h2>
          <ul>
            {providerObjs
              .sort(({ health: a }, { health: b }) => {
                console.log(Object.values(a));
                return (
                  Math.min(...Object.values(a)) - Math.min(...Object.values(b))
                );
              })
              .map(({ stop, health, available, settings }) => (
                <li>
                  <StopLine stop={stop} />
                  <table>
                    {Object.entries(health)
                      .sort(([, a], [, b]) => a - b)
                      .map(([item, health]) => (
                        <tr>
                          <td>
                            {(health * 100).toLocaleString('en', {
                              maximumFractionDigits: 0,
                            })}
                            %
                          </td>
                          <td>
                            <ItemIcon name={item} alt={item} />
                          </td>
                          <td>
                            <Item name={item} />
                          </td>
                        </tr>
                      ))}
                  </table>
                  {/*{JSON.stringify(available)}*/}
                  {/*{JSON.stringify(settings)}*/}
                </li>
              ))}
          </ul>
        </div>
        <div class="row">
          <h2>Names not matching a known pattern</h2>
          <ul>
            {bugs.map((stop) => (
              <li>
                <StopLine stop={stop} />
              </li>
            ))}
          </ul>
        </div>
        <div class="row">
          <h2>Names deemed too boring to consider</h2>
          <ul>
            {boring.map((stop) => (
              <li>
                <StopLine stop={stop} />
              </li>
            ))}
          </ul>
        </div>
      </>
    );
  }
}

class StopLine extends Component<{ stop: [string, Stop] }> {
  render(props: { stop: [string, Stop] }) {
    const stop = props.stop[1];
    const loc = props.stop[0];
    return (
      <span>
        <RenderIcons text={stop.name} /> (<a href={`/block/${loc}`}>{loc}</a>)
      </span>
    );
  }
}
