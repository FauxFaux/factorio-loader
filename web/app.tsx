import { Component } from 'preact';
import Router from 'preact-router';
import { createHashHistory } from 'history';

import { WhatTheBrick } from './pages/what-the-brick';
import { StationStatus } from './pages/station-status';
import { IoFDetail } from './pages/objects';
import { BlockPage } from './pages/block';
import { Map } from './pages/map';

import { LtnTree } from './ltn-tree';
import { ItemList, StationList } from './lists';

import 'leaflet/dist/leaflet.css';
import './main.css';
import { Next } from './pages/next';
import { Pulses } from './pages/pulses';
import { Chestify } from './pages/chestify';

export class App extends Component {
  render() {
    return (
      <div className="container">
        <div className="row">{header}</div>
        <Router history={createHashHistory() as any}>
          <Home path="/" />
          <WhatTheBrick path="/what-the-brick" />
          <StationStatus path="/station-status" />
          <IoFDetail path="/item/:name" type="item" name="from the path" />
          <IoFDetail path="/fluid/:name" type="fluid" name="from the path" />
          <BlockPage path="/block/:loc" loc="from the path" />
          <LtnTree
            path="/ltn-tree/:type/:name"
            type="item"
            name="from the path"
          />
          <Map
            path="/map/:gps?/:zoom?"
            gps="from the path"
            zoom="from the path"
          />
          <Next path="/an/next" />
          <Pulses path="/an/pulses/:colon" colon="from the path" />
          <Chestify path="/an/chestify" />
        </Router>
      </div>
    );
  }
}

const header = (
  <header className="p-3">
    <div className="container">
      <div className="d-flex flex-wrap align-items-center justify-content-center justify-content-lg-start">
        <ul className="nav col-12 col-lg-auto me-lg-auto mb-2 justify-content-center mb-md-0">
          <li>
            <a href="/" className="nav-link px-2 text-white">
              Station / Item search
            </a>
          </li>
          <li>
            <a href="/what-the-brick" className="nav-link px-2 text-white">
              What the brick?!
            </a>
          </li>
          <li>
            <a href="/station-status" className="nav-link px-2 text-white">
              Station status
            </a>
          </li>
          <li>
            <a href="/map" className="nav-link px-2 text-white">
              Map
            </a>
          </li>
          <li>
            <a href="/an/next" className="nav-link px-2 text-white">
              Next
            </a>
          </li>
          <li>
            <a
              href="https://github.com/FauxFaux/factorio-loader"
              className="nav-link px-2 text-white"
            >
              github
            </a>
          </li>
        </ul>
      </div>
    </div>
  </header>
);

class Home extends Component {
  render() {
    return (
      <div class="row">
        <div class="col">
          <h2>Stations</h2>
          <StationList limit={100} />
        </div>
        <div class="col">
          <h2>Items</h2>
          <ItemList limit={100} />
        </div>
      </div>
    );
  }
}
