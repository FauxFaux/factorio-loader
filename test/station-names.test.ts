import { provideStationPurpose } from '../web/station-status';
import { data } from '../web';
import { readFileSync } from 'fs';

for (const key of Object.keys(data)) {
  (data as any)[key] = JSON.parse(
    readFileSync(require.resolve(`../data/${key}.json`), { encoding: 'utf-8' }),
  );
}

test.each([
  '[item=duralumin] Duralumin Provide',
  'Iron Ore Provide',
  'Incubated Petri Dish Provide',
  'Useless Provide',
  'Auog Bonemeal Provide',
])('%s', (name) => {
  expect(provideStationPurpose(name)).toMatchSnapshot();
});
