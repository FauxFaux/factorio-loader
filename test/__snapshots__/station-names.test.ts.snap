// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[item=duralumin] Duralumin Provide 1`] = `
Set {
  "duralumin",
}
`;

exports[`Auog Bonemeal Provide 1`] = `
Set {
  "bonemeal",
}
`;

exports[`Chemical Science Pack Provide (from science!) 1`] = `
Set {
  "chemical-science-pack",
}
`;

exports[`Incubated Petri Dish Provide 1`] = `
Set {
  "petri-dish-bacteria",
}
`;

exports[`Iron Ore Provide 1`] = `
Set {
  "iron-ore",
}
`;

exports[`Laboratory Instruments Provide 1`] = `
Set {
  "lab-instrument",
}
`;

exports[`Useless Provide 1`] = `Set {}`;
